src/AbstractFactory.php

```diff
@@ -94,6 +94,8 @@ protected function retrieveConfig(ContainerInterface $container, string $configK
      * If the container does not know about the dependency, it is pulled from a fresh factory. This saves the user from
      * registering factories which they are not gonna access themself at all, and thus minimized configuration.
      *
+     * @param string|class-string $factoryClassName
```

Comment:
I think `class-string` is a Psalm only concept right? if so, could we make this `@psalm-param class-string`?

Reply:
`@param class-string<T> $name` is currently supported by Psalm, Phan and PHPStan. Not sure if having a generic docblock is better or pslam-only variants are just fine. I can simply follow the final decision, but just wanted to clarify:

 - Common docblocks (`@param`) are making things less noisy on PHPStorm side.
 - Having both is also an option but more noisy (`@param string` + `@psalm-param class-string`)
 - If method have only `@psalm-param class-string` without native type declaration, PHPStorm warns because of unknown|missing parameter type. Maybe they could also improve IDE inspections in the future but for today, its annoying.

Reply:
> PHPStorm warns because of unknown|missing parameter type.

Since `string $factoryClassName` is already the parameter type, as far as I can see the only complaint is missing PHPDoc, which is fine to ignore in this case:

![a44a9b5](https://user-images.githubusercontent.com/496145/74016164-dfa98f00-4989-11ea-891c-f47f4c66552c.png)

Since PHPStorm doesn't understand `class-string`, and Phpstan (not that we're using it here) does understand `@psalm-param`, I'd prefer just using `@psalm-param class-string $factoryClassName` and ignore the `Argument PHPDoc missing` inspection locally if it bothers anyone.

Reply:
But yes, I agree would be nice if PhpStorm supported more of Psalm's syntax :D maybe one day

Reply:
What I currently do is providing an `@param` for PHPStorm & co, and `@psalm-param` with the more specific type

Reply:
Ok I adapted `@pslam-` convention everywhere except StubFileDriver::loadMetadataForClass because it does not have native type declaration because of the hierarchy. I can not add native declaration there because of the hierarchy there. When we have `@pslam-` only hint, PHPCS starts warning because of missing both `@param` and native declarataion.

It's really hard keeping both analyzers happy. Now I am unhappy because the codebase mixed up with two different conventionsfor very same goal. Surpressing the PHPCS error using `ignore` may also an option, even if it's not cool.

Reply:
That's fine for now, thank you! :+1: